File: solo-leveling-player-service/src/main/java/com/sleepkqq/sololeveling/player/service/service/UserService.java
package com.sleepkqq.sololeveling.player.service.service;

import com.sleepkqq.sololeveling.player.service.model.User;
import com.sleepkqq.sololeveling.player.service.repository.UserRepository;
import java.util.Optional;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

@Slf4j
@Service
@RequiredArgsConstructor
public class UserService {

  private final UserRepository userRepository;

  public User get(long id) {
    return find(id).orElseThrow(() -> new IllegalArgumentException("User not found id=" + id));
  }

  public Optional<User> find(long id) {
    return userRepository.findById(id);
  }

  @Transactional
  public User createOrUpdate(User user) {
    userRepository.findById(user.getId())
        .ifPresent(existingUser -> {
          user.setVersion(existingUser.getVersion());
          user.setPlayer(existingUser.getPlayer());
        });
    return userRepository.save(user);
  }
}

File: solo-leveling-player-service/src/main/kotlin/com/sleepkqq/sololeveling/player/service/service/user/UserService.kt
package com.sleepkqq.sololeveling.player.service.service.user

import com.sleepkqq.sololeveling.player.model.entity.user.User
import com.sleepkqq.sololeveling.player.model.entity.user.dto.UserView
import java.time.LocalDateTime

interface UserService {
	fun get(id: Long): UserView
	fun find(id: Long): UserView?
	fun findVersion(id: Long): Int?
	fun insert(user: User): User
	fun update(user: User, now: LocalDateTime = LocalDateTime.now()): User
	fun upsert(user: User): User
}
